File: DKPlugins/DKDom/DKDomWebSocket.cpp
/*
* This source file is part of digitalknob, the cross-platform C/C++/Javascript/Html/Css Solution
*
* For the latest information, see https://github.com/aquawicket/DigitalKnob
*
* Copyright(c) 2010 - 2023 Digitalknob Team, and contributors
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files(the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and /or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions :
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/
#include "DK/DKApp.h"
#include "DKDom/DKDom.h"
#include "DKDom/DKDomWebSocket.h"
#include "DKUWebSocketsClient/DKUWebSocketsClient.h" //uwebsocket


bool DKDomWebSocket::Init() {
	DKDEBUGFUNC();
	
	//// Instance properties ////
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_binaryType",		DKDomWebSocket::binaryType);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_bufferedAmount",	DKDomWebSocket::bufferedAmount);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_extensions",		DKDomWebSocket::extensions);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_protocol",		DKDomWebSocket::protocol);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_readyState",		DKDomWebSocket::readyState);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_url",				DKDomWebSocket::url);
	
	//// Instance methods ////
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_open",			DKDomWebSocket::open);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_close",			DKDomWebSocket::close);
	DKDuktape::AttachFunction("CPP_DKDomWebSocket_send",			DKDomWebSocket::send);
	
	DKClass::DKCreate("DKDom/DKDomWebSocket.js");
	return true;
}

int DKDomWebSocket::binaryType(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomWebSocket::bufferedAmount(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomWebSocket::extensions(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomWebSocket::protocol(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomWebSocket::readyState(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomWebSocket::url(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomWebSocket::open(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	DKString address = duk_require_string(ctx, 0);
	if (!DKUWebSocketsClient::Get()->CreateClient(address)) //uwebsocket
		return DKERROR("DKWebSockets::CreateClient() failed! \n");
	return true;
}

int DKDomWebSocket::close(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	if (!DKUWebSocketsClient::Get()->CloseClient()) //uwebsocket
		return DKERROR("DKWebSockets::CloseClient() failed! \n");
	return true;
}

int DKDomWebSocket::send(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	DKString message = duk_require_string(ctx, 0);
	if (!DKUWebSocketsClient::Get()->MessageToServer(message)) //uwebsocket
		return DKERROR("DKWebSockets::MessageToServer() failed! \n");
	return true;
}

File: DKPlugins/DKDom/DKDomWebSocket.js
// [MDN] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket
// [INTERFACE] https://websockets.spec.whatwg.org/#the-websocket-interface


// [WebSocket()] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket
var WebSocket = function WebSocket(url) {
	console.log("WebSocket("+url+")");
	this.url = url;
	var pointer = CPP_DKDomDocument_createElement("websocket");
	this.pointer = pointer;
	CPP_DK_Create("DKUWebSocketsClient,"+pointer);
	CPP_DKDomWebSocket_open(url);

	////// Instance properties //////
	// [WebSocket.binaryType] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/binaryType
	Object.defineProperty(this, "binaryType", {
        get: function binaryType() {
            return CPP_DKDomWebSocket_binaryType(pointer)
        }
    })
	// [WebSocket.bufferedAmount](Read only) https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/bufferedAmount
	Object.defineProperty(this, "bufferedAmount", {
        get: function bufferedAmount() {
            return CPP_DKDomWebSocket_bufferedAmount(pointer)
        }
    })
	// [WebSocket.extensions](Read only) https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/extensions
	Object.defineProperty(this, "extensions", {
        get: function extensions() {
            return CPP_DKDomWebSocket_extensions(pointer)
        }
    })
	// [WebSocket.protocol](Read only) https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/protocol
	Object.defineProperty(this, "protocol", {
        get: function protocol() {
            return CPP_DKDomWebSocket_protocol(pointer)
        }
    })
	// [WebSocket.readyState](Read only) https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
	Object.defineProperty(this, "readyState", {
        get: function readyState() {
            return CPP_DKDomWebSocket_readyState(pointer)
        }
    })
	// [WebSocket.url](Read only) https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/url
	Object.defineProperty(this, "url", {
        get: function url() {
            return CPP_DKDomWebSocket_url(pointer)
        }
    })
	

	////// Instance methods //////
	WebSocket.prototype.close = function close() {
		CPP_DKDomWebSocket_close()
    }
	WebSocket.prototype.send = function send(message) {
		CPP_DKDomWebSocket_send(message)
    }
	
	
	//// Events ////
	// [close] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close_event
	// [error] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/error_event
	// [message] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/message_event
	// [open] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/open_event



	// NOTE: already included from DKDomGlobalEventHandlers.js
	/*
	this.close_func = null;
    Object.defineProperty(this, "onclose", {
        get: function onclose() {
            return this.close_func;
        },
        set: function onclose(func) {
            func && this.addEventListener("close", func)
			!func && this.removeEventListener("close", this.close_func)
            this.close_func = func
        }
    });
	*/
	
	// NOTE: already included from DKDomGlobalEventHandlers.js
	/*
	this.error_func = null;
    Object.defineProperty(this, "onerror", {
        get: function onerror() {
            return this.error_func;
        },
        set: function onerror(func) {
            func && this.addEventListener("error", func)
			!func && this.removeEventListener("error", this.error_func)
            this.error_func = func
        }
    });
	*/

	this.message_func = null;
    Object.defineProperty(this, "onmessage", {
        get: function onmessage() {
            return this.message_func;
        },
        set: function onmessage(func) {
            func && this.addEventListener("message", func)
			!func && this.removeEventListener("message", this.message_func)
            this.message_func = func
        }
    });

	this.open_func = null;
    Object.defineProperty(this, "onopen", {
        get: function onopen() {
            return this.open_func;
        },
        set: function onopen(func) {
            func && this.addEventListener("open", func)
			!func && this.removeEventListener("open", this.open_func)
            this.open_func = func
        }
    });

	
	//// toString ////
	if(this.toString() === "[object Object]")
		this.toString = function(){	return "[object WebSocket]" }
	
    return EventTarget.call(this, pointer)
}
WebSocket.prototype = EventTarget.prototype;